### PR TITLE
Revert "HIVE-23725: ValidTxnManager snapshot outdating causing partia…

### PR DESCRIPTION
**Pull request by deniskuzZ**
…l reads (Peter Varga, reviewed by Jesus Camacho Rodriguez, Denys Kuzmenko)"

This reverts commit e2a02f1b43cba657d4d1c16ead091072be5fe834.

### What changes were proposed in this pull request?

reverts https://issues.apache.org/jira/browse/HIVE-23725

### Why are the changes needed?

doesn't completely solve the problem described in a JIRA, will be replaced with another solution

### Does this PR introduce _any_ user-facing change?

No

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/reexec/ReExecutionDagSubmitPlugin.java [hunk 1]
```diff
@@ -76,7 +76,7 @@ public void beforeExecute(int executionIndex, boolean explainReOptimization) {
   }
 
   @Override
-  public boolean shouldReExecute(int executionNum, CommandProcessorException ex) {
+  public boolean shouldReExecute(int executionNum) {
     return (executionNum < maxExecutions) && retryPossible;
```
Comment (kgyrtkirk) [inline ql/src/java/org/apache/hadoop/hive/ql/reexec/ReExecutionDagSubmitPlugin.java:80]: please change this to `retryPossible` ( and remove maxExecutions field)

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java [hunk 2]
```diff
@@ -2324,142 +2315,6 @@ private void testConcurrentMergeInsertNoDuplicates(String query, boolean sharedW
...
-  }
-
-  /**
-   * ValidTxnManager.isValidTxnListState can invalidate a snapshot if a relevant write transaction was committed
-   * between a query compilation and lock acquisition. When this happens we have to recompile the given query,
-   * otherwise we can miss reading partitions created between. The following three cases test these scenarios.
-   * @throws Exception ex
-   */
-  @Test
-  public void testMergeInsertDynamicPartitioningSequential() throws Exception {
```
Comment (pvargacl) [inline ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:2337]: These tests should be added back in your next change, when we recompile without the reexec driver, to verify the dyn partitioning use-cases

Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:2337]: Ack.

File: ql/src/java/org/apache/hadoop/hive/ql/reexec/ReExecutionDagSubmitPlugin.java (hunk 1 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/reexec/ReExecutionDagSubmitPlugin.java]: changed

File: standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift [hunk 3]
```diff
@@ -1012,8 +1012,9 @@ struct CommitTxnRequest {
     3: optional list<WriteEventInfo> writeEventInfos,
     // Information to update the last repl id of table/partition along with commit txn (replication from 2.6 to 3.0)
     4: optional ReplLastIdInfo replLastIdInfo,
+    5: optional bool exclWriteEnabled = true,
```
Comment (pvargacl) [inline standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift:1015]: This should be added as the last parameter to not break backward compatibility no?

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java [hunk 4]
```diff
@@ -2315,6 +2386,139 @@ private void testConcurrentMergeInsertNoDuplicates(String query, boolean sharedW
...
+    Assert.assertEquals("Duplicate records found", 6, res.size());
+    Assert.assertTrue("Partition 3 was skipped", res.contains("6\t6\t3"));
+    dropTable(new String[]{"target", "source"});
+  }
+
+  @Test
+  public void testMergeInsertDynamicPartitioningSnapshotInvalidatedWithOldCommit() throws Exception {
+    // By creating the driver with the factory, we should have a ReExecDriver
+    IDriver driver3 = DriverFactory.newDriver(conf);
+    Assert.assertTrue("ReExecDriver was expected", driver3 instanceof ReExecDriver);
```
Comment (pvargacl) [inline ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java:2432]: This Reexec part is not really relevant now, we don't need the reexec driver for this to work properly

File: ql/src/java/org/apache/hadoop/hive/ql/Driver.java [hunk 5]
```diff
@@ -488,30 +489,40 @@ private void runInternal(String command, boolean alreadyCompiled) throws Command
...
           driverContext.getConf().set(ValidTxnList.VALID_TXNS_KEY,
             driverContext.getTxnManager().getValidTxns().toString());
+
           if (driverContext.getPlan().hasAcidResourcesInQuery()) {
+            compileInternal(context.getCmd(), true);
             driverTxnHandler.recordValidWriteIds();
+            driverTxnHandler.setWriteIdForAcidFileSinks();
           }
 
           if (!alreadyCompiled) {
```
Comment (pvargacl) [inline ql/src/java/org/apache/hadoop/hive/ql/Driver.java:528]: I think this code should be removed. If the alreadyCompiled was false, we already compiled it once line 473. This should not matter when we are in the invalid snapshot case, you already recompile the query if it is neccessarry

Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/Driver.java]: removed

File: ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java (hunk 4 above)
Comment (deniskuzZ) [reply ql/src/test/org/apache/hadoop/hive/ql/lockmgr/TestDbTxnManager2.java]: changed

File: standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift (hunk 3 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-common/src/main/thrift/hive_metastore.thrift:1015]: we need to keep it consistent with downstream, see `HIVE-23759: refactor field order of CommitTxnRequest`

**Comment by pvargacl** [comment]
LGTM +1

**Review by pvary — APPROVED** [review]

**Pull request merged**